### PR TITLE
chore(swingset): add slogging framework

### PR DESCRIPTION
**Pull request by warner**
This adds a set of logging hooks which record swingset deliveries (the kernel
sending messages and resolution notifications into vats), syscalls (from the
vat into the kernel), and console messages. All these events can be
correlated by their crank number, vatID, delivery number, and syscall number.

The output function is not yet implemented; my notional plan is to log to
lines of JSON for now, and maybe move to a SQLite based store for later. I've
not written any analysis tools yet either.

I made a few refactorings to make things smoother:

* vats get their console object from `managerOptions.vatConsole`, rather than
  having the manager build it themselves, which makes it easier to provide
  one that writes to both the regular console and the slog
* as a result, all vats get the same basic `vatEndowments` (with the console
  added on later), rather than calling `makeVatEndowments(vatID)`
* in kernel.js, `deliverToVat` and `processNotify` had their common code
  factored out into `deliverAndLogToVat()`, which now writes to the slog
  in addition to actually performing the `manager.deliver`. The names aren't
  great and want to be changed at some point.
* `buildVatSyscallHandler` function was moved out of `addVatManager` to make
  things easier to read. The new function writes syscalls to the slog.
* console messages written by vats in top-level code, or during the execution
  of `buildRootObject`, are slogged in the "startup" phase. All other calls
  to `console.log` and friends are recorded in a "delivery" phase.

refs #1535

### CODE REVIEW

File: packages/SwingSet/src/kernel/kernel.js
Comment (FUDCo) [inline packages/SwingSet/src/kernel/kernel.js:280]: This name is really confusing.  My first thought was that it should be `deliverToVatAndLog` but "log" is also a noun so it's still ambiguous.  Really what you mean (I think) is `deliverToVatThenLog` but that's still a little ambiguous and anyway is not a name anyone should ever use.  Or you could say `deliverToVatAndThenLogDelivery` but now we're into COBOL levels of foolishness.  Perhaps another approach is to declare that relevant logging is implicit in everything we do and so just call it `deliverToVat`.  But the bottom line is this needs a better name.

File: packages/SwingSet/src/kernel/slogger.js
Comment (FUDCo) [inline packages/SwingSet/src/kernel/slogger.js:35]: I think I fundamentally don't buy in to the traditional hierarchy of logging levels, but this is a longer conversation to have before we land this.

File: packages/SwingSet/src/kernel/slogger.js
Comment (FUDCo) [inline packages/SwingSet/src/kernel/slogger.js:57]: I don't understand logging both `kd` and `vd`.  Since one is a translation of the other, I would think that the only reason to log them both is if you are debugging the translation mechanism itself, which is unusual now and will be almost unheard of later.  Doing it universally and perpetually seems like a lot of overhead to carry, especially in production.

File: packages/SwingSet/src/kernel/slogger.js
Comment (FUDCo) [inline packages/SwingSet/src/kernel/slogger.js:67]: I'm not wild about this idiom of having each of the log functions return a finish function.  What do we know at the beginning that we don't know at the end?

**Review by FUDCo — COMMENTED** [review]
I'm not in love with the flavor this is developing -- it's not a "this needs these specific changes before approval" thing, it's more a generalized anxiety that this is looking to turn into a very heavyweight thing that feels awkward and expensive -- something we should probably have a more extended design conversation about before implementing a lot more mechanism.

File: packages/SwingSet/src/kernel/kernel.js
Comment (warner) [reply packages/SwingSet/src/kernel/kernel.js:280]: Yeah, as mentioned in the commit comment, these three functions (deliverToVat, processNotify, and deliverAndLogToVat) want to have a name refactoring. I think this particular function wants to be named `deliverToVat`, because that's exacly what it's doing. The previous deliverToVat is only dealing with message sends (not notifies), so it wants a name that includes "message" or "send" or something.

I think a deeper issue is that I've been inconsistent (and drifting) around my naming for these delivery things. I started out using "dispatch" as the verb/noun, as in `dispatch.deliver` and `dispatch.notifyFulfillToPresence`, but 1: it's harder to say than just "deliver", and 2: as this kernel-to-vat invocation moves to being more about sending a data object from kernel to vatManager to vatWorker (and less about invoking some particular method of a `dispatch` object, which now happens inside the vatWorker, maybe inside a different process altogether, and generally invisible to the kernel), I want a name for the data object. "Dispatch" is both a verb and a noun (as in "dispatches from the eastern front"), but "delivery" has the benefit of being obviously not a verb. It may help to hash out a new naming system for these pieces, and then rename all the functions to match.

File: packages/SwingSet/src/kernel/slogger.js
Comment (warner) [reply packages/SwingSet/src/kernel/slogger.js:35]: Yeah, the goal here is simply to match the functionality of the original `console` object. This code is actually copied from the `anylogger` wrapper that @michaelfig put together, which we might want to augment with the more exotic `console` methods (there are some timestamping ones that we omitted).

File: packages/SwingSet/src/kernel/slogger.js
Comment (warner) [reply packages/SwingSet/src/kernel/slogger.js:57]: My goal is to give the rendering tool (which has access to the slog data, but not the HostDB, which might be far away or long-gone by now, e.g. a dynamic vat might have been deleted) enough data to correlate cross-vat messages, as well as correlate deliveries/syscalls with log entries created by e.g. liveslots. So I need at least the `slots` in both kernel form and vat form.

It'd be sufficient (more work here, less data stored in the slog file) to store just the KernelDelivery record, but also parse the record to find all the slots (message target, message result, message argument slots, promise resolution target, promise resolution presence, promise resolution/rejection data slots), translate them (again) (and without accidentally allocating anything) (and without re-retiring anything), and record a table mapping all those slots to their vat-side equivalents.

I figured it was easier and less dangerous (in the face of retiring promises) to just record both the pre- and post- translation records.

File: packages/SwingSet/src/kernel/slogger.js
Comment (warner) [reply packages/SwingSet/src/kernel/slogger.js:67]: Yeah, I'm on the fence about that approach myself.

The reasons we tell the slog about both delivery-start and delivery-end is so 1: we can record the delivery results somewhere ("the vat died", or "metering reports 14 computrons consumed"), and 2: so the slog knows we're done with a delivery, because any syscalls that happen after that point are illegal (the vat is supposed to be idle).

The delivery results need to be correlated with the delivery start record, and they are matched with a `deliveryNum` (specifically the tuple of `vatID` and `deliveryNum`). But `deliveryNum` is hidden inside the slog code: the kernel doesn't keep track of them. (`crankNum` is recorded too, to let the rendering tool show a total ordering of deliveries, but I'm hesitant to use it for delivery start/finish correlation purposes.. maybe I shouldnt' be). So what `finish` is really closing over is that particular `deliveryNum` value.

But we could implement that by having the `start` return `deliveryNum` instead of `finish()`, and expose a separate `kernelSlog.finishDelivery(vatID, deliveryNum, deliveryResult)` method. The extra arguments seemed redundant to me, an opportunity to get them wrong.

(an earlier iteration exposed distinct `vatSlog` objects for each vat, to remove the `vatID` argument from most of the calls, but finding a place to store it and still access it from the right places seemed messy.. maybe I should revisit that)

File: packages/SwingSet/src/kernel/kernel.js
Comment (warner) [reply packages/SwingSet/src/kernel/kernel.js]: Aye. I think I can remove that flag (and the `kernelOptions` entirely, for now) by switching on the presence of the `writeSlogObject` kernel endowment instead. I only added it to support a pair of unit tests (`test-kernel.js` and I think `test-vpid-kernel.js`) which `buildKernel` without the controller.

One negative consequence of the "return a `finish()` function" you raised elsewhere is that building a mock/dummy `kernelSlog` object is somewhat less trivial: it has to return dummy `finish()` functions too. If it returned an object which was supposed to be passed back into a separate `finish` method, then the dummy slog would be simpler. I think there's a TC39 proposal to add `?` to the language in a way that would make this even better (`const logthing = kernelSlog?.startDelivery(..)` followed by `kernelSlog?.finishDelivery(logthing, ..)`).

Comment (FUDCo) [reply packages/SwingSet/src/kernel/kernel.js]: I actually kinda like the "kernel options" idea in principle -- there have been a few times in the course of developing swingset runner that I wanted something like that, to control how the kernel gets set up in principled, non-ad hoc way.

**Review by FUDCo — APPROVED** [review]
....

**Pull request merged**
